hnn_core/network.py

```diff
@@ -219,47 +245,86 @@ def __repr__(self):
-                                                     self.params['tstop'])
-        self._n_common_feeds = len(self._p_common)
+        # XXX CHECKS
+        # 1. drive_name not already present
 
         # 'position' the artificial cells arbitrarily in the origin of the
         # network grid. Non-biophysical cell placement is irrelevant
         # However, they must be present to be included in gid_ranges!
         origin = self.pos_dict['origin']
 
-        # COMMON FEEDS
-        self.pos_dict['common'] = [origin for i in range(self._n_common_feeds)]
-
-        # UNIQUE FEEDS
-        for key in self._p_unique.keys():
-            # create the pos_dict for all the sources
-            self.pos_dict[key] = [origin for i in range(self.n_cells)]
+        if cell_specific:
+            drive_cell_range = range(self.n_cells)
+        else:
+            drive_cell_range = range(len(drive_params))
+        # create the pos_dict for all the sources
+        self.pos_dict[drive_name] = [origin for i in drive_cell_range]
 
-        # Now add the names of the feeds to a list
-        self.feedname_list.append('common')
-        # grab the keys for the unique set of inputs and sort the names
-        # append them to the src list along with the number of cells
-        unique_keys = sorted(self._p_unique.keys())
-        self.feedname_list += unique_keys
+        # NB must be added before updating gid_ranges!
+        self.external_drives.update({drive_name: list()})
 
         # Every time pos_dict is updated, gid_ranges must be updated too
         self._update_gid_ranges()
 
-        # Create the feed dynamics (event_times)
-        self._instantiate_feeds(n_trials=self.params['N_trials'])
-
-    def _instantiate_feeds(self, n_trials=1):
-        """Creates event_time vectors for all feeds and all trials
+        drive_dict = {'drive_name': drive_name, 'params': drive_params,
+                      'target_cell_type': None, 'gid': None}
+
+        if cell_specific:
+            for gid in self.gid_ranges[drive_name]:
+                gid_target = gid - self.gid_ranges[drive_name][0]
+                target_cell_type = self.gid_to_type(gid_target)
+                drive_dict['target_cell_type'] = target_cell_type
+                drive_dict['params'] = drive_params  # for parconnect
+                drive_dict['gid'] = gid
+                self.external_drives[drive_name].append(deepcopy(drive_dict))
+        else:
+            # XXX this for loop is needed to support legacy-HNN 'common feeds'
+            for idx, gid in enumerate(self.gid_ranges[drive_name]):
+                drive_dict['target_cell_type'] = None  # attaches to all
+                drive_dict['params'] = drive_params[idx]
+                drive_dict['gid'] = gid
+                self.external_drives[drive_name].append(deepcopy(drive_dict))
+                self._n_common_feeds += 1
+
+    def attach_bias(self, bias_name):
```

Comment:
I don't get this ... why do you make `_attach_tonic_bias` private? We need to expose those parameters to the user. Since there is only tonic bias as of now, I'd suggest sticking to a single method rather than trying to build a more generic wrapper. Over-engineering is evil

Reply:
In this case, I agree. The rationale for 'hiding' anything is to adhere to the design goals of HNN: it heavily restricts the user. I have no problems making adding tonic biases explicit.